File: ParsFramework/Util/Main/Pars.Util.Puma/Pars.Util.Puma.Domain/Range.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;

namespace Pars.Util.Puma.Domain
{
    [DataContract]
    public class Range<T>
    {
        public static Range<T> From(T from) =>
            new Range<T>(from, default(T));

        public static Range<T> Until(T until) =>
            new Range<T>(default(T), until);

        public static Range<T> Between(T from, T until) =>
            new Range<T>(from, until);

        public bool IsFrom => Start != null && End == null;
        public bool IsUntil => Start == null && End != null;
        public bool IsBetween => Start != null && End != null;
        public bool IsNoRange => Start == null && End == null;

        [DataMember(Name = "start")]
        public T Start { get; set; }
        [DataMember(Name = "end")]
        public T End { get; set; }

        public Range(T start, T end)
        {
            Start = start;
            End = end;
        }

        public Range() : this(default(T), default(T))
        {
        }
    }

}

File: ParsFramework/Util/Main/Pars.Util.Puma/Pars.Util.Puma.UI.Mvc/app/pars/data/range.ts

/**
 * 
 */
export class Range<T> {

    /**
     * 
     * @param from
     */
    public static from<T>(from: T): Range<T> {
        return new Range(from, null);
    }

    /**
     * 
     * @param until
     */
    public static until<T>(until: T): Range<T> {
        return new Range(null, until);
    }

    public static between<T>(from: T, until: T): Range<T> {
        return new Range(from, until);
    }

    /**
     * 
     * @returns {} 
     */
    get isFrom(): boolean {
        return this.start != null && this.end == null;
    }

    /**
     * 
     * @returns {} 
     */
    get isUntil(): boolean {
        return this.start == null && this.end != null;
    }

    /**
     * 
     * @returns {} 
     */
    get isBetween(): boolean {
        return this.start != null && this.end != null;
    }

    /**
     * 
     * @returns {} 
     */
    get isNoRange(): boolean {
        return this.start == null && this.end == null;
    }

    constructor(public start: T, public end: T) { }

}


